File: leetcode/problems/cpp/subsets.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(1)

#include "vector"
using namespace std;

class Solution {
public:
   vector<vector<int>> subsets(vector<int>& nums) {
      int size = nums.size();
      int mask = 1 << size;
      vector<vector<int>> subsets(mask);
      for (int i = 1; i < mask; i++)
         for (int j = 0;j < size; j++)
            if (i & (1 << j))
               subsets[i].push_back(nums[j]);
      return subsets;
   }
};

File: leetcode/problems/java/subsets.java
import java.util.List;
import java.util.ArrayList;

class Solution {
  public List<List<Integer>> subsets(int[] nums) {
    List<List<Integer>> res = new ArrayList<>();
    backtrack(res, new ArrayList<>(), nums, 0);
    return res;
  }

  void backtrack(List<List<Integer>> res, List<Integer> temp, int[] nums, int start) {
    if (start >= nums.length) {
      res.add(new ArrayList<>(temp));
      return;
    }
    temp.add(nums[start]);
    backtrack(res, temp, nums, start + 1);
    temp.remove(temp.size() - 1);
    backtrack(res, temp, nums, start + 1);
  }
}